src/goExtractMethod.ts

```diff
@@ -0,0 +1,172 @@
+ *
+ * @param methodName name for the extracted method
+ * @param selection the editor selection from which method is to be extracted
+ * @param editor the editor that will be used to apply the changes from godoctor
+ * @returns errorMessage in case the method fails, null otherwise
+ */
+export function extractMethodUsingGoDoctor(methodName: string, selection: Selection, editor: TextEditor): Thenable<string> {
+	let godoctor = getBinPath('godoctor');
+	let position = `${selection.start.line + 1},${selection.start.character + 1}:${selection.end.line + 1},${selection.end.character + 1}`;
+
+	return new Promise((resolve, reject) => {
+		let process = cp.execFile(godoctor, ['-pos', position, 'extract', methodName], {}, (err, stdout, stderr) => {
+			if (err) {
+				let errorMessageIndex = stderr.indexOf('Error:');
+				return resolve(errorMessageIndex > -1 ? stderr.substr(errorMessageIndex) : stderr);
+			}
+
+			let d = new dmp.diff_match_patch();
+			let patchText = stdout.substr(stdout.indexOf('@@'));
+			let patches: dmp.Patch[];
+
+			try {
+				patches = d.patch_fromText(patchText);
+			}
+			catch (e) {
+				return resolve(`Failed to parse the patches from godoctor: ${e.message}`);
+			}
+
+			applypatches(patches, editor).then(validEdit => {
+				return resolve (validEdit ? null : 'Edits could not be applied to the document');
+			});
+
+		});
+		process.stdin.end(editor.document.getText());
+	});
+}
+
+/**
+ * Applies the given set of patches to the document in the given editor
+ *
+ * @param patches array of patches to be applied
+ * @param editor the TextEditor whose document will be updated
+ */
+function applypatches(patches: dmp.Patch[], editor: TextEditor): Thenable<boolean> {
+	let totalEdits: Edit[] = [];
+	patches.reverse().forEach((patch: dmp.Patch) => {
+		let edits = getEditsFromPatch(patch);
+		totalEdits = totalEdits.concat(edits);
+	});
+	let editPromise = editor.edit((editBuilder) => {
+		totalEdits.forEach((edit) => {
+			switch (edit.action) {
+				case EditTypes.EDIT_INSERT:
+					editBuilder.insert(edit.start, edit.text);
+					break;
+				case EditTypes.EDIT_DELETE:
+					editBuilder.delete(new Range(edit.start, edit.end));
+					break;
+				case EditTypes.EDIT_REPLACE:
+					editBuilder.replace(new Range(edit.start, edit.end), edit.text);
+					break;
+			}
+		});
+	});
+
+	// While inserting the extracted method, the end of line char (\n) at the end of the document might get replaced.
+	// If this happens, the next time godoctor is called, it adds "No End of Line" char which will fail diff_match_patch.patch_fromtext()
+	// Therefore, add the end of line char at the end of the document if it does not exists.
+	return editPromise.then(done => {
+		if (done && !editor.document.getText().endsWith('\n')) {
+			return editor.edit((editBuilder) => {
+				editBuilder.insert(new Position(editor.document.lineCount, 0), '\n');
```

Comment:
@Tyriar Extracted methods get added at the end of the document. If `\n` gets added to the input, then the resulting diffs/patches will assume this `\n` exists and not output one. This will result in the extracted method getting appended to the last line instead of starting from a new line _after_ the last line.